File: app/src/main/java/com/example/phonebook/views/adapters/ContactAdapter.java
package com.example.phonebook.views.adapters;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.RecyclerView;

import com.example.phonebook.databinding.ContactViewBinding;
import com.example.phonebook.model.data.Contact;
import com.example.phonebook.model.service.ContactClickCallback;

import java.util.ArrayList;

public class ContactAdapter extends RecyclerView.Adapter<ContactAdapter.ContactViewHolder>{

    @Nullable
    private final ContactClickCallback mContactClickCallback;

    private ArrayList<Contact> mContacts;

    public ContactAdapter (@Nullable ContactClickCallback _callBack) {
        mContactClickCallback = _callBack;
    }

    public void updateContactList(ArrayList<Contact> _contacts) {
        if(mContacts == null) {
            mContacts = _contacts;
            notifyDataSetChanged();
        } else {
            DiffUtil.DiffResult result = DiffUtil.calculateDiff(new DiffUtil.Callback() {

                @Override
                public int getOldListSize() {
                    return mContacts.size();
                }

                @Override
                public int getNewListSize() {
                    return _contacts.size();
                }

                @Override
                public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
                    return mContacts.get(oldItemPosition).getID().equals(_contacts.get(newItemPosition).getID());
                }

                @Override
                public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
                    return mContacts.get(oldItemPosition).equals(_contacts.get(newItemPosition));
                }
            });

            mContacts = _contacts;
            result.dispatchUpdatesTo(this);
        }
    }

    @NonNull
    @Override
    public ContactViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        ContactViewBinding binding = ContactViewBinding.inflate(LayoutInflater.from(parent.getContext()), parent, false);
        binding.setCallback(mContactClickCallback);
        return new ContactViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull ContactViewHolder holder, int position) {
        holder.binding.setContact(mContacts.get(position));
        holder.binding.executePendingBindings();
    }

    @Override
    public int getItemCount() {
        return mContacts == null ? 0 : mContacts.size();
    }

    static class ContactViewHolder extends RecyclerView.ViewHolder{

        private final ContactViewBinding binding;

        ContactViewHolder (ContactViewBinding _binding) {
            super(_binding.getRoot());
            this.binding = _binding;
        }
    }
}

File: app/src/main/java/com/example/phonebook/views/adapters/ContactAdapter.kt
package com.example.phonebook.views.adapters

import android.view.LayoutInflater
import android.view.ViewGroup

import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView

import com.example.phonebook.databinding.ContactViewBinding
import com.example.phonebook.model.data.Contact
import com.example.phonebook.model.service.ContactClickCallback

import java.util.ArrayList

class ContactAdapter(callBack: ContactClickCallback) : RecyclerView.Adapter<ContactAdapter.ContactViewHolder>() {

    private val mContactClickCallback: ContactClickCallback = callBack

    private var mContacts: ArrayList<Contact>? = null

    fun updateContactList(contacts: ArrayList<Contact>? ) {
        if(mContacts == null) {
            mContacts = contacts
            notifyDataSetChanged()
        } else {
            val result: DiffUtil.DiffResult  = DiffUtil.calculateDiff(object: DiffUtil.Callback() {

                override fun getOldListSize() = mContacts!!.size

                override fun getNewListSize() = contacts!!.size

                override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int) = mContacts!![oldItemPosition].ID == contacts!![newItemPosition].ID

                override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int) = mContacts!![oldItemPosition] == contacts!![newItemPosition]
            })

            result.dispatchUpdatesTo(this)
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ContactViewHolder {
        val binding: ContactViewBinding = ContactViewBinding.inflate(LayoutInflater.from(parent.context), parent, false).apply {
            callback = mContactClickCallback
        }

        return ContactViewHolder(binding)
    }

    @Override
    override fun onBindViewHolder(holder: ContactViewHolder, position: Int) {
        holder.binding.contact = mContacts?.get(position)
        holder.binding.executePendingBindings()
    }

    override fun getItemCount(): Int {
        return mContacts?.size ?: 0
    }

    inner class ContactViewHolder(val binding: ContactViewBinding) : RecyclerView.ViewHolder(binding.root)
}
